File: deployment_package/backend/core/raha_regime_integration.py
"""
RAHA Market Regime Integration
Applies market regime analysis to RAHA strategy execution and position sizing
"""
import logging
from typing import Dict, Any, Optional, Tuple
from decimal import Decimal
from .rust_stock_service import rust_stock_service
from .raha_models import RAHASignal, StrategyVersion

logger = logging.getLogger(__name__)


class RAHARegimeIntegration:
    """
    Integrates market regime oracle with RAHA strategy execution.
    Provides regime-aware position sizing and risk controls.
    """
    
    # Strategy type mapping from RAHA strategy names to regime strategy types
    STRATEGY_MAPPING = {
        'ORB': 'ORB_MOMENTUM',
        'MOMENTUM': 'TREND_SWING',
        'BREAKOUT': 'ORB_MOMENTUM',
        'PULLBACK': 'PULLBACK_BUY',
        'MEAN_REVERSION': 'MEAN_REVERSION_INTRADAY',
        'SWING': 'TREND_SWING',
    }
    
    def __init__(self):
        self.rust_service = rust_stock_service
    
    def get_regime_analysis(
        self,
        symbol: str,
        secondary: Optional[str] = None
    ) -> Optional[Dict[str, Any]]:
        """
        Get market regime analysis for a symbol.
        
        Args:
            symbol: Primary symbol to analyze
            secondary: Secondary symbol for correlation (defaults to SPY)
        
        Returns:
            Dict with regime analysis or None if unavailable
        """
        try:
            response = self.rust_service.analyze_correlation(symbol, secondary)
            
            if not response:
                logger.warning(f"No regime analysis available for {symbol}")
                return None
            
            return {
                'global_regime': response.get('global_regime', 'NEUTRAL'),
                'local_context': response.get('local_context', 'NORMAL'),
                'correlation_1d': response.get('correlation_1d', 0.0),
                'correlation_7d': response.get('correlation_7d', 0.0),
                'correlation_30d': response.get('correlation_30d', 0.0),
                'btc_dominance': response.get('btc_dominance'),
                'spy_correlation': response.get('spy_correlation'),
            }
        except Exception as e:
            logger.warning(f"Error getting regime analysis for {symbol}: {e}")
            return None
    
    def get_position_multiplier(
        self,
        signal: RAHASignal,
        regime_analysis: Optional[Dict[str, Any]] = None
    ) -> Decimal:
        """
        Calculate position size multiplier based on regime.
        
        Args:
            signal: RAHA signal
            regime_analysis: Pre-fetched regime analysis (optional)
        
        Returns:
            Multiplier to apply to base position size (1.0 = no change)
        """
        if not regime_analysis:
            regime_analysis = self.get_regime_analysis(signal.symbol)
        
        if not regime_analysis:
            # No regime data available, use neutral multiplier
            return Decimal('1.0')
        
        global_regime = regime_analysis.get('global_regime', 'NEUTRAL')
        local_context = regime_analysis.get('local_context', 'NORMAL')
        
        # Map RAHA strategy to regime strategy type
        strategy_name = signal.strategy_version.strategy.name.upper() if signal.strategy_version else 'UNKNOWN'
        strategy_type = self.STRATEGY_MAPPING.get(strategy_name, 'TREND_SWING')
        
        # Get multiplier from Rust service (or use hardcoded defaults)
        # For now, we'll use hardcoded multipliers based on the regime matrix
        multiplier = self._get_regime_multiplier(strategy_type, global_regime, local_context)
        
        logger.info(
            f"Regime multiplier for {signal.symbol} ({strategy_type}): {multiplier}x "
            f"(regime: {global_regime}, context: {local_context})"
        )
        
        return Decimal(str(multiplier))
    
    def _get_regime_multiplier(
        self,
        strategy_type: str,
        global_regime: str,
        local_context: str
    ) -> float:
        """
        Get position size multiplier based on strategy type and regime.
        This matches the Rust raha_regime_integration.rs matrix.
        """
        # Base multipliers by regime
        base_multipliers = {
            'EQUITY_RISK_ON': {
                'ORB_MOMENTUM': 1.3,
                'TREND_SWING': 1.4,
                'MEAN_REVERSION_INTRADAY': 0.9,
                'PULLBACK_BUY': 1.3,
            },
            'EQUITY_RISK_OFF': {
                'ORB_MOMENTUM': 0.7,
                'TREND_SWING': 0.6,
                'MEAN_REVERSION_INTRADAY': 0.8,
                'PULLBACK_BUY': 0.5,
            },
            'CRYPTO_ALT_SEASON': {
                'ORB_MOMENTUM': 1.1,
                'TREND_SWING': 1.2,
                'MEAN_REVERSION_INTRADAY': 0.9,
                'PULLBACK_BUY': 1.1,
            },
            'CRYPTO_BTC_DOMINANCE': {
                'ORB_MOMENTUM': 0.8,
                'TREND_SWING': 0.7,
                'MEAN_REVERSION_INTRADAY': 0.9,
                'PULLBACK_BUY': 0.7,
            },
            'NEUTRAL': {
                'ORB_MOMENTUM': 1.0,
                'TREND_SWING': 1.0,
                'MEAN_REVERSION_INTRADAY': 1.0,
                'PULLBACK_BUY': 1.0,
            },
        }
        
        # Get base multiplier
        base = base_multipliers.get(global_regime, {}).get(strategy_type, 1.0)
        
        # Adjust for local context
        if local_context == 'IDIOSYNCRATIC_BREAKOUT':
            base *= 1.1  # Slight bonus for decoupled moves
        elif local_context == 'CHOPPY_MEAN_REVERT':
            base *= 0.9  # Reduce size in choppy markets
        
        return base
    
    def get_risk_controls(
        self,
        symbol: str,
        regime_analysis: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        """
        Get risk control adjustments based on regime.
        
        Returns:
            Dict with risk control parameters:
            - stop_loss_tightness: Multiplier for stop loss (1.0 = normal, <1.0 = tighter, >1.0 = wider)
            - take_profit_aggressiveness: Multiplier for take profit (1.0 = normal, <1.0 = take earlier, >1.0 = let run)
            - max_trades_multiplier: Multiplier for max trades per day
        """
        if not regime_analysis:
            regime_analysis = self.get_regime_analysis(symbol)
        
        if not regime_analysis:
            return {
                'stop_loss_tightness': 1.0,
                'take_profit_aggressiveness': 1.0,
                'max_trades_multiplier': 1.0,
            }
        
        global_regime = regime_analysis.get('global_regime', 'NEUTRAL')
        
        # Risk controls by regime (matching Rust implementation)
        controls = {
            'EQUITY_RISK_ON': {
                'stop_loss_tightness': 0.8,  # Wider stops (let winners breathe)
                'take_profit_aggressiveness': 1.2,  # Let winners run
                'max_trades_multiplier': 1.1,
            },
            'EQUITY_RISK_OFF': {
                'stop_loss_tightness': 1.3,  # Tighter stops
                'take_profit_aggressiveness': 0.7,  # Take profits earlier (scalp)
                'max_trades_multiplier': 0.7,
            },
            'CRYPTO_ALT_SEASON': {
                'stop_loss_tightness': 0.9,
                'take_profit_aggressiveness': 1.1,
                'max_trades_multiplier': 1.0,
            },
            'CRYPTO_BTC_DOMINANCE': {
                'stop_loss_tightness': 1.2,
                'take_profit_aggressiveness': 0.8,
                'max_trades_multiplier': 0.8,
            },
            'NEUTRAL': {
                'stop_loss_tightness': 1.0,
                'take_profit_aggressiveness': 1.0,
                'max_trades_multiplier': 1.0,
            },
        }
        
        return controls.get(global_regime, controls['NEUTRAL'])
    
    def get_regime_narration(
        self,
        symbol: str,
        regime_analysis: Optional[Dict[str, Any]] = None
    ) -> Optional[Dict[str, str]]:
        """
        Get human-readable narration for the current regime.
        
        Returns:
            Dict with title, description, and action_summary
        """
        if not regime_analysis:
            regime_analysis = self.get_regime_analysis(symbol)
        
        if not regime_analysis:
            return None
        
        global_regime = regime_analysis.get('global_regime', 'NEUTRAL')
        
        narrations = {
            'EQUITY_RISK_ON': {
                'title': '📈 Equity RISK-ON',
                'description': 'Markets are in a RISK-ON equity regime. Trends are being rewarded.',
                'action_summary': "We're slightly increasing size on high-conviction continuation setups and giving winners more room.",
            },
            'EQUITY_RISK_OFF': {
                'title': '⚠️ Equity RISK-OFF',
                'description': "We're in a RISK-OFF equity regime. Volatility is elevated and correlations to the index are breaking down.",
                'action_summary': "I'm reducing your size, tightening stops, and favouring defined-risk or defensive setups.",
            },
            'CRYPTO_ALT_SEASON': {
                'title': '🌕 Alt-Season',
                'description': 'Crypto is in ALT-SEASON. BTC dominance is falling and high-quality alts are moving in sync with the broader risk trend.',
                'action_summary': "I'm slightly increasing size in your highest-ranked alts and rotating out of weaker names.",
            },
            'CRYPTO_BTC_DOMINANCE': {
                'title': '🟠 BTC-Dominance',
                'description': "We're in a BTC-DOMINANCE regime. Capital is hiding in BTC rather than chasing alts.",
                'action_summary': "I'm shrinking your alt exposure and focusing on higher-quality majors with defined risk.",
            },
            'NEUTRAL': {
                'title': '⚪ Neutral',
                'description': 'Market regime is neutral. No strong directional bias detected.',
                'action_summary': 'Using standard position sizing and risk controls.',
            },
        }
        
        return narrations.get(global_regime, narrations['NEUTRAL'])


# Global instance
raha_regime_integration = RAHARegimeIntegration()

File: rust_crypto_engine/src/raha_regime_integration.rs
use serde::{Serialize, Deserialize};
use crate::correlation_analysis::{GlobalRegime, LocalContext};

/// Strategy types that RAHA can execute
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum RahaStrategy {
    OrbMomentum,           // Opening Range Breakout momentum
    TrendSwing,            // Trend-following swing trades
    MeanReversionIntraday, // Mean-reversion intraday plays
    PullbackBuy,           // Pullback buys in strong names
    ShortWeakSectors,      // Short-side plays in weak sectors
    DefinedRiskOptions,    // Defined-risk options spreads
    BtcEthSwing,           // BTC/ETH swing trades
    AltTrendFollowing,     // Trend-following on strong alts
    RotationalTrade,      // Rotate from laggards to leaders
}

/// Position sizing multiplier based on regime
/// Base size is multiplied by this value
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PositionSizingMultiplier {
    pub strategy: RahaStrategy,
    pub equity_risk_on: f64,
    pub equity_risk_off: f64,
    pub crypto_alt_season: f64,
    pub crypto_btc_dominance: f64,
    pub neutral: f64,
    pub choppy_mean_revert: f64,
}

/// Risk controls that change based on regime
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RegimeRiskControls {
    pub max_position_size_multiplier: f64,
    pub stop_loss_tightness: f64, // 1.0 = normal, <1.0 = tighter, >1.0 = wider
    pub take_profit_aggressiveness: f64, // 1.0 = normal, <1.0 = take profits earlier, >1.0 = let winners run
    pub max_trades_per_day_multiplier: f64,
    pub allow_aggressive_fades: bool,
}

/// UX narration text for each regime
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RegimeNarration {
    pub regime: GlobalRegime,
    pub title: String,
    pub description: String,
    pub action_summary: String,
}

/// RAHA regime integration engine
/// Maps market regimes to RAHA behavior (position sizing, strategy selection, risk controls)
pub struct RahaRegimeIntegration {
    strategy_multipliers: Vec<PositionSizingMultiplier>,
}

impl RahaRegimeIntegration {
    pub fn new() -> Self {
        Self {
            strategy_multipliers: Self::default_strategy_matrix(),
        }
    }

    /// Get position sizing multiplier for a strategy in a given regime
    pub fn get_position_multiplier(
        &self,
        strategy: &RahaStrategy,
        global_regime: &GlobalRegime,
        local_context: &LocalContext,
    ) -> f64 {
        let base_multiplier = self
            .strategy_multipliers
            .iter()
            .find(|m| &m.strategy == strategy)
            .map(|m| match global_regime {
                GlobalRegime::EquityRiskOn => m.equity_risk_on,
                GlobalRegime::EquityRiskOff => m.equity_risk_off,
                GlobalRegime::CryptoAltSeason => m.crypto_alt_season,
                GlobalRegime::CryptoBtcDominance => m.crypto_btc_dominance,
                GlobalRegime::Neutral => m.neutral,
            })
            .unwrap_or(1.0);

        // Adjust for local context
        let context_multiplier = match local_context {
            LocalContext::IdiosyncraticBreakout => 1.1, // Slight bonus for decoupled moves
            LocalContext::ChoppyMeanRevert => 0.9,      // Reduce size in choppy markets
            LocalContext::Normal => 1.0,
        };

        base_multiplier * context_multiplier
    }

    /// Get risk controls for a given regime
    pub fn get_risk_controls(&self, global_regime: &GlobalRegime) -> RegimeRiskControls {
        match global_regime {
            GlobalRegime::EquityRiskOn => RegimeRiskControls {
                max_position_size_multiplier: 1.3,
                stop_loss_tightness: 0.8, // Wider stops (let winners breathe)
                take_profit_aggressiveness: 1.2, // Let winners run
                max_trades_per_day_multiplier: 1.1,
                allow_aggressive_fades: false,
            },
            GlobalRegime::EquityRiskOff => RegimeRiskControls {
                max_position_size_multiplier: 0.6,
                stop_loss_tightness: 1.3, // Tighter stops
                take_profit_aggressiveness: 0.7, // Take profits earlier (scalp mentality)
                max_trades_per_day_multiplier: 0.7,
                allow_aggressive_fades: false,
            },
            GlobalRegime::CryptoAltSeason => RegimeRiskControls {
                max_position_size_multiplier: 1.2,
                stop_loss_tightness: 0.9,
                take_profit_aggressiveness: 1.1,
                max_trades_per_day_multiplier: 1.0,
                allow_aggressive_fades: false,
            },
            GlobalRegime::CryptoBtcDominance => RegimeRiskControls {
                max_position_size_multiplier: 0.5, // Shrink alt exposure
                stop_loss_tightness: 1.2,
                take_profit_aggressiveness: 0.8,
                max_trades_per_day_multiplier: 0.8,
                allow_aggressive_fades: false,
            },
            GlobalRegime::Neutral => RegimeRiskControls {
                max_position_size_multiplier: 1.0,
                stop_loss_tightness: 1.0,
                take_profit_aggressiveness: 1.0,
                max_trades_per_day_multiplier: 1.0,
                allow_aggressive_fades: true,
            },
        }
    }

    /// Get UX narration for a regime
    pub fn get_narration(&self, global_regime: &GlobalRegime) -> RegimeNarration {
        match global_regime {
            GlobalRegime::EquityRiskOn => RegimeNarration {
                regime: GlobalRegime::EquityRiskOn,
                title: "📈 Equity RISK-ON".to_string(),
                description: "Markets are in a RISK-ON equity regime. Trends are being rewarded.".to_string(),
                action_summary: "We're slightly increasing size on high-conviction continuation setups and giving winners more room.".to_string(),
            },
            GlobalRegime::EquityRiskOff => RegimeNarration {
                regime: GlobalRegime::EquityRiskOff,
                title: "⚠️ Equity RISK-OFF".to_string(),
                description: "We're in a RISK-OFF equity regime. Volatility is elevated and correlations to the index are breaking down.".to_string(),
                action_summary: "I'm reducing your size, tightening stops, and favouring defined-risk or defensive setups.".to_string(),
            },
            GlobalRegime::CryptoAltSeason => RegimeNarration {
                regime: GlobalRegime::CryptoAltSeason,
                title: "🌕 Alt-Season".to_string(),
                description: "Crypto is in ALT-SEASON. BTC dominance is falling and high-quality alts are moving in sync with the broader risk trend.".to_string(),
                action_summary: "I'm slightly increasing size in your highest-ranked alts and rotating out of weaker names.".to_string(),
            },
            GlobalRegime::CryptoBtcDominance => RegimeNarration {
                regime: GlobalRegime::CryptoBtcDominance,
                title: "🟠 BTC-Dominance".to_string(),
                description: "We're in a BTC-DOMINANCE regime. Capital is hiding in BTC rather than chasing alts.".to_string(),
                action_summary: "I'm shrinking your alt exposure and focusing on higher-quality majors with defined risk.".to_string(),
            },
            GlobalRegime::Neutral => RegimeNarration {
                regime: GlobalRegime::Neutral,
                title: "⚪ Neutral".to_string(),
                description: "Market regime is neutral. No strong directional bias detected.".to_string(),
                action_summary: "Using standard position sizing and risk controls.".to_string(),
            },
        }
    }

    /// Get recommended strategies for a regime (prioritized list)
    pub fn get_recommended_strategies(&self, global_regime: &GlobalRegime) -> Vec<RahaStrategy> {
        match global_regime {
            GlobalRegime::EquityRiskOn => vec![
                RahaStrategy::OrbMomentum,
                RahaStrategy::TrendSwing,
                RahaStrategy::PullbackBuy,
            ],
            GlobalRegime::EquityRiskOff => vec![
                RahaStrategy::ShortWeakSectors,
                RahaStrategy::DefinedRiskOptions,
                RahaStrategy::MeanReversionIntraday,
            ],
            GlobalRegime::CryptoAltSeason => vec![
                RahaStrategy::AltTrendFollowing,
                RahaStrategy::RotationalTrade,
                RahaStrategy::TrendSwing,
            ],
            GlobalRegime::CryptoBtcDominance => vec![
                RahaStrategy::BtcEthSwing,
                RahaStrategy::DefinedRiskOptions,
            ],
            GlobalRegime::Neutral => vec![
                RahaStrategy::MeanReversionIntraday,
                RahaStrategy::OrbMomentum,
                RahaStrategy::TrendSwing,
            ],
        }
    }

    /// Default strategy × regime matrix
    fn default_strategy_matrix() -> Vec<PositionSizingMultiplier> {
        vec![
            PositionSizingMultiplier {
                strategy: RahaStrategy::OrbMomentum,
                equity_risk_on: 1.3,
                equity_risk_off: 0.7,
                crypto_alt_season: 1.1,
                crypto_btc_dominance: 0.8,
                neutral: 1.0,
                choppy_mean_revert: 0.9,
            },
            PositionSizingMultiplier {
                strategy: RahaStrategy::TrendSwing,
                equity_risk_on: 1.4,
                equity_risk_off: 0.6,
                crypto_alt_season: 1.2,
                crypto_btc_dominance: 0.7,
                neutral: 1.0,
                choppy_mean_revert: 0.9,
            },
            PositionSizingMultiplier {
                strategy: RahaStrategy::MeanReversionIntraday,
                equity_risk_on: 0.9,
                equity_risk_off: 0.8,
                crypto_alt_season: 0.9,
                crypto_btc_dominance: 0.9,
                neutral: 1.0,
                choppy_mean_revert: 1.2,
            },
            PositionSizingMultiplier {
                strategy: RahaStrategy::PullbackBuy,
                equity_risk_on: 1.3,
                equity_risk_off: 0.5,
                crypto_alt_season: 1.1,
                crypto_btc_dominance: 0.7,
                neutral: 1.0,
                choppy_mean_revert: 0.8,
            },
            PositionSizingMultiplier {
                strategy: RahaStrategy::ShortWeakSectors,
                equity_risk_on: 0.3,
                equity_risk_off: 1.2,
                crypto_alt_season: 0.5,
                crypto_btc_dominance: 0.5,
                neutral: 0.8,
                choppy_mean_revert: 0.9,
            },
            PositionSizingMultiplier {
                strategy: RahaStrategy::DefinedRiskOptions,
                equity_risk_on: 0.8,
                equity_risk_off: 1.1,
                crypto_alt_season: 0.9,
                crypto_btc_dominance: 1.0,
                neutral: 1.0,
                choppy_mean_revert: 1.0,
            },
            PositionSizingMultiplier {
                strategy: RahaStrategy::BtcEthSwing,
                equity_risk_on: 1.0,
                equity_risk_off: 0.8,
                crypto_alt_season: 0.7,
                crypto_btc_dominance: 1.1,
                neutral: 1.0,
                choppy_mean_revert: 0.9,
            },
            PositionSizingMultiplier {
                strategy: RahaStrategy::AltTrendFollowing,
                equity_risk_on: 1.0,
                equity_risk_off: 0.6,
                crypto_alt_season: 1.3,
                crypto_btc_dominance: 0.5,
                neutral: 1.0,
                choppy_mean_revert: 0.8,
            },
            PositionSizingMultiplier {
                strategy: RahaStrategy::RotationalTrade,
                equity_risk_on: 1.1,
                equity_risk_off: 0.7,
                crypto_alt_season: 1.2,
                crypto_btc_dominance: 0.6,
                neutral: 1.0,
                choppy_mean_revert: 0.9,
            },
        ]
    }
}

impl Default for RahaRegimeIntegration {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_position_multipliers() {
        let integration = RahaRegimeIntegration::new();
        
        // ORB momentum should be larger in risk-on
        let mult = integration.get_position_multiplier(
            &RahaStrategy::OrbMomentum,
            &GlobalRegime::EquityRiskOn,
            &LocalContext::Normal,
        );
        assert_eq!(mult, 1.3);
        
        // ORB momentum should be smaller in risk-off
        let mult = integration.get_position_multiplier(
            &RahaStrategy::OrbMomentum,
            &GlobalRegime::EquityRiskOff,
            &LocalContext::Normal,
        );
        assert_eq!(mult, 0.7);
        
        // Mean reversion should be larger in choppy markets
        let mult = integration.get_position_multiplier(
            &RahaStrategy::MeanReversionIntraday,
            &GlobalRegime::Neutral,
            &LocalContext::ChoppyMeanRevert,
        );
        assert_eq!(mult, 1.2);
    }

    #[test]
    fn test_risk_controls() {
        let integration = RahaRegimeIntegration::new();
        
        let controls = integration.get_risk_controls(&GlobalRegime::EquityRiskOn);
        assert_eq!(controls.max_position_size_multiplier, 1.3);
        assert_eq!(controls.stop_loss_tightness, 0.8); // Wider stops
        
        let controls = integration.get_risk_controls(&GlobalRegime::EquityRiskOff);
        assert_eq!(controls.max_position_size_multiplier, 0.6);
        assert_eq!(controls.stop_loss_tightness, 1.3); // Tighter stops
    }
}

